### PR TITLE
fix(serve): support serve hmr

### PR DESCRIPTION
**Pull request by HasanAlrimawi**
This PR addresses issue #25600 

Changes:
Updated `fn has_hmr` to check `serve` subcommand and return its hmr value if found, in order to run the worker in serve mode with hmr_runner. Thus the hmr event can be dispatched upon changes on the file served.

### CODE REVIEW

File: cli/worker.rs [hunk 1]
```diff
@@ -415,7 +414,7 @@ impl CliMainWorker {
 
 #[derive(Clone)]
 pub struct CliMainWorkerFactory {
-  shared: Arc<SharedWorkerState>,
+  pub shared: Arc<SharedWorkerState>,
```
Comment (nathanwhit) [inline cli/worker.rs:417]: Are these visibility changes needed?

Comment (HasanAlrimawi) [reply cli/worker.rs:417]: No not at all, I used that while inspecting, I'll remove it soon.

File: cli/tools/serve.rs [hunk 2]
```diff
@@ -44,12 +44,18 @@ pub async fn serve(
   maybe_npm_install(&factory).await?;
 
   let worker_factory = factory.create_cli_main_worker_factory().await?;
-
+  let hmr = serve_flags
+    .watch
+    .unwrap_or(WatchFlagsWithPaths {
+      hmr: false,
+      ..Default::default()
+    })
```
Comment (bartlomieju) [inline cli/tools/serve.rs:52]:
```suggestion
.map(|watch_flags| watch_flags.hmr).unwrap_or(false)
```

File: cli/worker.rs [hunk 3]
```diff
@@ -121,7 +121,6 @@ pub struct CliMainWorkerOptions {
   pub serve_port: Option<u16>,
   pub serve_host: Option<String>,
 }
-
```
Comment (bartlomieju) [inline cli/worker.rs:124]: Can you please revert this line?

**Review by bartlomieju — COMMENTED** [review]
@nathanwhit please take a look. I'm not sure this actually helps much - looks like HMR runner is still not wired up properly

**Comment by HasanAlrimawi** [comment]
@nathanwhit 
Could you please give it a look and let me know if any further work is needed?

File: cli/worker.rs (hunk 3 above)
Comment (nathanwhit) [reply cli/worker.rs:127]: Looks like this hasn't been reverted yet

**Comment by nathanwhit** [comment]
Aside from that [one whitespace change](https://github.com/denoland/deno/pull/26078#discussion_r1821362394) that should be reverted, LGTM

**Review by nathanwhit — APPROVED** [review]
LGTM, thanks!

**Pull request merged**